File: vectors.py
"""
Vector Module

This module defines a Vector class for performing vector operations.
It provides methods for creating vectors from magnitude and angles, calculating magnitudes,
performing dot and cross products, checking for vector equality and negativity, and more.

Classes:
    Vector: A class representing a vector and providing various vector operations.

Author: Lokesh Kaushik
Date: August 22, 2023
"""

import math


class Vector:

    def __init__(self, i=None, j=None, k=None):
        self.i = i if i is not None else 0
        self.j = j if j is not None else 0
        self.k = k if k is not None else 0

    @classmethod
    def from_magnitude(cls, magnitude, angle_xy, angle_xz=None) :
        ''' takes magnitude and angle of xy plane and xz plane, then returns a vector with x, y, z coordinates'''
        x = round(magnitude * math.cos(math.radians(angle_xy)), 2)
        y = round(magnitude * math.sin(math.radians(angle_xy)), 2)

        if angle_xz is not None:
            z = round(magnitude * math.cos(math.radians(angle_xz)), 2)
            return cls(x, y, z)
        else:
            return cls(x, y, 0)

    @classmethod
    def angle_between(cls, vector1, vector2):
        '''take 2 vectors then calculate the angle between them and returns the angle which is of float type'''
        product = vector1.dot_product(vector2)
        magnitude_product = vector1.magnitude() * vector2.magnitude()

        cos_angle = product / magnitude_product
        angle_radians = math.acos(cos_angle)
        angle_degrees = round(math.degrees(angle_radians), 2)

        return angle_degrees


    @classmethod
    def is_negative(cls,first,other):
        ''' returns 1 if one vector is negate of other'''
        if(first.to_negative() == other):
            return 1
        else:
            return 0
        

    @classmethod
    def is_equal(cls,first,other):
        ''' return 1 if both vectors have same magnitude and direction'''
        if(first == other):
            return 1
        else:
            return 0
        

    def display(self):
        ''' returns nothing but display x, y, z of a vector'''
        print(f"{self.i}i, {self.j}j, {self.k}k")

    def magnitude(self):
        ''' calculate the magnitude of a vector and returns a float value'''
        return math.sqrt(self.i ** 2 + self.j ** 2 + self.k ** 2)

    def dot_product(self, other):
        '''performs dot product on two vectors and returns a value of type int or float depends upon given x, y, z'''
        product = self.i * other.i + self.j * other.j + self.k * other.k

        return product

    def crossProduct(self, other):
        ''' performs cross product on two vectors and returns a new vector'''
        new_i = self.j * other.k - self.k * other.j
        new_j = self.k * other.i - self.i * other.k
        new_k = self.i * other.j - self.j * other.i

        return Vector(new_i, new_j, new_k)

    def to_negative(self):
        ''' return negate of a vector'''
        x = self.i * -1
        y = self.j * -1
        z = self.k * -1

        return Vector(x, y, z)
    

    def __add__(self, other):
        new_i = self.i + other.i
        new_j = self.j + other.j
        new_k = self.k + other.k

        return Vector(new_i, new_j, new_k)

    def __sub__(self, other):
        new_i = self.i - other.i
        new_j = self.j - other.j
        new_k = self.k - other.k

        return Vector(new_i, new_j, new_k)

    def __mul__(self, other):
        if isinstance(other, (int, float)):
            new_i = self.i * other
            new_j = self.j * other
            new_k = self.k * other

            return Vector(new_i, new_j, new_k)
        
    def __eq__(self, other):
        return self.i == other.i and self.j == other.j and self.k == other.k
    
    def __str__(self):
        return f"{self.i}i, {self.j}j, {self.k}k"


if __name__ == "__main__":
   v1 = Vector(3,5,6)
   v2 = Vector(2,8,4)
   print(v1+v2)

File: vectors.cpp
#include <iostream>
#include <math.h>
using namespace std;

class Vector {
    float i, j, k;

public:
    // Constructor with default values (0.0) for x, y, and z components
    Vector(float x = 0.0, float y = 0.0, float z = 0.0) : i(x), j(y), k(z) {}

    // Display the vector components
    void display() {
        cout << i << "i, " << j << "j, " << k << "k" << endl;
    }

    // Calculate the magnitude of the vector
    float magnitude() {
        return sqrt(pow(i, 2) + pow(j, 2) + pow(k, 2));
    }

    // Calculate the dot product with another vector
    float dot_product(const Vector other) {
        return i * other.i + j * other.j + k * other.k;
    }

    // Create a vector from magnitude and angles (in degrees)
    static Vector from_magnitude(float mag, float angle_xy, float angle_xz = 0.0) {
        float x, y, z;

        x = mag * cos(angle_xy * (M_PI / 180.0));
        y = mag * sin(angle_xy * (M_PI / 180.0));

        if (angle_xz != 0.0)
            z = mag * cos(angle_xz * (M_PI / 180.0));
        else
            z = 0;

        return Vector(round(x * 100.0) / 100.0, round(y * 100.0) / 100.0, round(z * 100.0) / 100.0);
    }

    // Calculate the angle between two vectors (in degrees)
    static float angle_between(Vector vec1, Vector vec2) {
        float vector_product = vec1.dot_product(vec2);
        float magnitude_product = vec1.magnitude() * vec2.magnitude();
        float angle_in_degree = (acos(vector_product / magnitude_product) * 180.0 / M_PI);

        return round(angle_in_degree * 100.0) / 100.0;
    }

    // Check if two vectors are negative of each other
    static bool is_negative(Vector vec1, const Vector vec2) {
        if (vec1.to_negative() == vec2)
            return true;
        return false;
    }

    // Check if two vectors are equal
    static bool is_equal(const Vector vec1, const Vector vec2) {
        if (vec1 == vec2)
            return true;
        return false;
    }

    // Calculate the cross product with another vector
    Vector cross_product(Vector other) {
        float new_i = j * other.k - k * other.j;
        float new_j = k * other.i - i * other.k;
        float new_k = i * other.j - j * other.i;

        return Vector(new_i, new_j, new_k);
    }

    // Return the negative of the current vector
    Vector to_negative() {
        return Vector(i * -1, j * -1, k * -1);
    }

    // Operator definitions

    // Check if two vectors are equal
    bool operator==(const Vector &other) const {
        return (i == other.i) && (j == other.j) && (k == other.k);
    }

    // Add two vectors
    Vector operator+(const Vector &other) const {
        return Vector(i + other.i, j + other.j, k + other.k);
    }

    // Subtract two vectors
    Vector operator-(const Vector &other) const {
        return Vector(i - other.i, j - other.j, k - other.k);
    }

    // Multiply a vector by a scalar
    Vector operator*(const float &other) const {
        return Vector(i * other, j * other, k * other);
    }

    // Output stream operator overload
    friend std::ostream &operator<<(std::ostream &os, const Vector vec) {
        os << vec.i << "i, " << vec.j << "j, " << vec.k << "k";
        return os;
    }

    // Input stream operator overload
    friend std::istream &operator>>(std::istream &is, Vector &vec) {
        cout << "i: ";
        is >> vec.i;
        cout << "j: ";
        is >> vec.j;
        cout << "k: ";
        is >> vec.k;
        return is;
    }
};

int main() {
    Vector a, b;
    cin >> a;
    cout << a + (a * 2);  // Display the result of a + (a * 2)
    return 0;
}
